Add support for image loading progress

tvolkert opened the pull request:
## Description

Add support for image loading progress.

This PR adds a `imageLoadingBuilder` property to the `Image` class.  This is useful for images that are incrementally loaded (e.g. over a local file system or a network) and you want to give the user an indication of when the image will be displayed.

## Related Issues

Fixes #32374

## Tests

I added the following tests:

* `consolidateHttpClientResponseBytes` notifies onBytesReceived for every chunk of bytes
* `NetworkImage` notifies listeners of chunk events
* `ImageStreamCompleter` delivers chunk events to listeners
* `ImageStreamCompleter` drops chunks events on the floor (does not buffer the events) if there are no listeners
* `ImageStreamCompleter` reports errors in the chunk event stream
* `Image` invokes `imageLoadingBuilder` as the image loads
* `Image` ignores chunk events if `imageLoadingBuilder` is null

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

tvolkert committed:
Add support for image loading progress

Fixes #32374

gspencergoog commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -6,19 +6,48 @@ import 'dart:async';
+/// The `total` parameter will contain the _expected_ total number of bytes to
+/// be received (extracted from the value of the `Content-Length` HTTP response
+/// header), or -1 if the size of the response body is not known in advance.
+/// Even if the parameter reports a non-negative value, it is not 100%
```
The `Content-Length` is the size of the transferred data, whether or not it is compressed, at least according to the [spec](https://tools.ietf.org/html/rfc2616#page-119). So this should be a good proxy for the progress, and I'm not sure that the comment is needed.  Do ask @Hixie, however, because he probably knows if the spec is implemented properly by most servers.

dnfield replied:
I had added this comment to fix a bug previously.

Some servers in the wild won't set this properly.  And ones that do set it properly will give you the compressed size, but dart:io automagically decompresses for you without telling you whether it's compressed.  At one point I had patched this so that it would try to be smart and detect that, but that was abandoned since the header may not even be reliable.

dnfield replied:
It may be worth documenting here that `total` could be less than `cumulative` when GZIP is used.

Do you think we could use the content-type header to determine if the response is one of our supported image formats or a likely compression  type?

tvolkert replied:
I updated the code to better handle gzipped content.  Now, `consolidateHttpClientResponseBytes()` will unzip the raw bytes itself if needed, which allowed us to turn off `autoUncompress` in the http client used by `NetworkImage`.

I also updated the docs.

gspencergoog commented on packages/flutter/test/painting/image_provider_test.dart:
```diff
@@ -15,19 +16,23 @@ import '../rendering/rendering_tester.dart';
 import 'mocks_for_image_cache.dart';
 
+/// A valid encoding of a 1x1 GIF.
+final Uint8List _smallGif = Uint8List.fromList(<int>[
```
Don't add your own, just use one from [image_data.dart](https://github.com/flutter/flutter/blob/master/packages/flutter/test/painting/image_data.dart)

tvolkert replied:
Done

tvolkert commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -6,19 +6,48 @@ import 'dart:async';
+/// Signature for getting notified when chunks of bytes are received while
+/// consolidating the bytes of an [HttpClientResponse] into a [Uint8List].
+///
+/// The `cumulative` parameter will contain the total number of bytes received
```
I tried to at first, but the analyzer didn't pick them up as the same - it doesn't seem to work on typedefs like it does in methods.

FYI @bwilkerson

tvolkert commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -6,19 +6,48 @@ import 'dart:async';
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far.
+///
+/// The `total` parameter will contain the _expected_ total number of bytes to
```
ditto

tvolkert committed:
Review comments

tvolkert committed:
Finish adding tests

dnfield commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -6,19 +6,48 @@ import 'dart:async';
+  // to the stream, so we manually count the bytes as they cross the wire.
   final Completer<Uint8List> completer = Completer<Uint8List>.sync();
   final List<List<int>> chunks = <List<int>>[];
+  final int expectedContentLength = response.contentLength;
```
This will almost always be less than the actual content length - your progress bar will end up loading faster because the contentLength will usually be the gzip compressed length.  You should be able to notice this if you test a network download of this with something from Wikimedia for example.

tvolkert replied:
Fixed.

tvolkert committed:
Make consolidateHttpClientResponseBytes() capable of uncompressing
the bytes manually, rather than relying on HttpClient.autoUncompress.

Correspondingly, change NetworkImage's HttpClient to turn off
autoUncompress, so as to get a more valid reading of the expected
content length.

(also, some minor doc updates)

tvolkert commented:
@gspencergoog @dnfield PTAL

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -68,14 +68,59 @@ class ImageInfo {
 
-class _ImageListenerPair {
-  _ImageListenerPair(this.listener, this.errorListener);
+/// An immutable notification of image bytes having been incrementally loaded.
```
```suggestion
/// An immutable notification of image bytes that have been incrementally loaded.
```

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -144,6 +190,14 @@ class ImageStream extends Diagnosticable {
   /// then use this flag to avoid calling [RenderObject.markNeedsPaint] during
   /// a paint.
   ///
+  /// An [ImageChunkListener] can also optionally be added along with the
```
```suggestion
  /// An [ImageChunkListener] can optionally be added along with the
```

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -144,6 +190,14 @@ class ImageStream extends Diagnosticable {
+  /// (e.g. over a file system or a network), then the specified listener will
+  /// be notified for every chunk of data that is loaded. The events that are
+  /// delivered to the listener are _not_ buffered (rather, they are fired and
+  /// forgotten), so it'd possible that the listener will have missed delivery
```
```suggestion
  /// forgotten), so it's possible that the listener will have missed delivery
```

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -266,6 +324,14 @@ abstract class ImageStreamCompleter extends Diagnosticable {
   /// then use this flag to avoid calling [RenderObject.markNeedsPaint] during
   /// a paint.
   ///
+  /// An [ImageChunkListener] can also optionally be added along with the
```
```suggestion
  /// An [ImageChunkListener] can optionally be added along with the
```

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -266,6 +324,14 @@ abstract class ImageStreamCompleter extends Diagnosticable {
+  /// (e.g. over a file system or a network), then the specified listener will
+  /// be notified for every chunk of data that is loaded. The events that are
+  /// delivered to the listener are _not_ buffered (rather, they are fired and
+  /// forgotten), so it'd possible that the listener will have missed delivery
```
```suggestion
  /// forgotten), so it's possible that the listener will have missed delivery
```

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -266,6 +324,14 @@ abstract class ImageStreamCompleter extends Diagnosticable {
+  /// be notified for every chunk of data that is loaded. The events that are
+  /// delivered to the listener are _not_ buffered (rather, they are fired and
+  /// forgotten), so it'd possible that the listener will have missed delivery
+  /// of some [ImageChunkEvent]s by the time they add themselves as a listener.
```
```suggestion
  /// of some [ImageChunkEvent]s by the time it is added as a listener.
```
(no need to anthropomorphize)

gspencergoog commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -500,11 +570,18 @@ class MultiFrameImageStreamCompleter extends ImageStreamCompleter {
   /// [scale] is the linear scale factor for drawing this frames of this image
   /// at their intended size.
+  ///
+  /// [chunkEvents] is an optional stream of notifications about the loading
```
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-correct-grammar
```suggestion
  /// The [chunkEvents] parameter is an optional stream of notifications about the loading
```

gspencergoog commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -29,6 +29,37 @@ export 'package:flutter/painting.dart' show
+/// progress.
+///
+/// This is useful for images that are incrementally loaded (e.g. over a local
+/// file system or a network) and you want to give the user an indication of
```
Avoid 'you'.

tvolkert replied:
I can change it, but pourquoi (I see 152 files in flutter/lib and flutter_test/lib) with "you" in the comment.

tvolkert replied:
Either way, done 😄

gspencergoog commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -566,6 +602,52 @@ class Image extends StatefulWidget {
+  ///
+  /// Once an image has fully loaded, this builder will no longer be consulted.
+  ///
+  /// {@tool sample}
```
You could make this a snippet pretty easily...

tvolkert replied:
Done

tvolkert committed:
Apply suggestions from code review

Co-Authored-By: Greg Spencer <gspencergoog@users.noreply.github.com>

tvolkert committed:
Convert sample to snippet

dnfield reviewed (approved):
LGTM

tvolkert committed:
Review comments

tvolkert committed:
Changed `client` param to consolidateHttpClientResponseBytes to be an optional named parameter.

It turns out a lot of code outside Flutter is using that method, so it needs to be a non-breaking change to the signature.

tvolkert committed:
Self-review

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+///
+/// The `cumulative` parameter will contain the total number of bytes received
+/// thus far. If the response has been gzipped, this number will be the number
+/// of compressed bytes that have received _across the wire_.
```
`number of (probably) compressed bytes`

tvolkert replied:
Why probably?  We're running it through a gzip decoded, so if it wasn't properly compressed, this method will throw.

Hixie replied:
The transfer might not have been compressed at all.

tvolkert replied:
But I preface the statement with "If the response has been gzipped".  I don't think I'm understanding what you're saying.

Hixie replied:
Oh I see. I misread it, sorry.

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+/// Efficiently converts the response body of an [HttpClientResponse] into a
+/// [Uint8List].
 ///
 /// The future returned will forward all errors emitted by [response].
```
it can only forward one error... maybe s/all errors/any error/?

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
-  // to the stream.
+///
+/// The [onBytesReceived] callback, if specified, will be invoked for every
+/// chunk of bytes that are received while consolidating the response bytes.
```
s/are/is/

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+Future<Uint8List> consolidateHttpClientResponseBytes(
+  HttpClientResponse response, {
+  HttpClient client,
+  BytesReceivedCallback onBytesReceived,
```
see our style guide for naming guidelines regarding the onBytesReceived argument

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+      expectedContentLength = -1;
+    } else {
+      // We need to un-compress the bytes as they come in.
+      sink = gzip.decoder.startChunkedConversion(output);
```
why do we want to override autoUncompress?

tvolkert replied:
A search yielded no usages of people calling this method with a client that has `autoUncompress` off, so we have a chance to define the behavior, so I just defined it as "this method will always return uncompressed bytes".  The use case is setting the client to not auto-uncompress so that we can get an accurate (or at least, _more likely to be accurate_) accounting of the expected content length.  When the client has auto-uncompressed, we have no chance of knowing the expected content length.

An alternative would be to add a new `autoUncompress` parameter to this method so callers could control the behavior (it'd be a no-op if the client had already uncompressed).  But since it looked like no one was using this with clients that kept the response bytes compressed, it seemed simpler to just define the method as always uncompressing.

Hixie replied:
How do you download a gzip with this API?

tvolkert replied:
heh fair enough - I'll add a parameter 😄

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+  final _OutputBuffer output = _OutputBuffer();
+  ByteConversionSink sink = output;
+  int expectedContentLength = response.contentLength;
+  if (response.headers?.value(HttpHeaders.contentEncodingHeader) == 'gzip') {
```
what if it's "identity, gzip"?

tvolkert replied:
I wondered about making this more robust, but I opted for copying the exact behavior from `HttpClient` so as to not add confusion as to how we determine that unzipping is necessary.

Hixie replied:
Please at a minimum file a bug and add a TODO

tvolkert replied:
According to [the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec3.html#sec3.5), "identity" is an invalid value in the Content-Encoding header.

> identity The default (identity) encoding; the use of no transformation whatsoever. This content-coding is used only in the Accept- Encoding header, and SHOULD NOT be used in the Content-Encoding header.

So I think the current impl here is correct.

Hixie commented on packages/flutter/lib/src/foundation/consolidate_response.dart:
```diff
@@ -3,31 +3,103 @@
+    sink.add(chunk);
+    if (onBytesReceived != null) {
+      bytesReceived += chunk.length;
+      onBytesReceived(bytesReceived, expectedContentLength);
```
should we document that exceptions from this callback will be uncaught and will cause the future never to complete?

Hixie replied:
(actually i guess they will be caught by `listen`?)

Hixie replied:
i guess since i'm confused about this, we should document it, whatever the behaviour is.

Hixie commented on packages/flutter/lib/src/painting/debug.dart:
```diff
@@ -31,3 +31,13 @@ bool debugAssertAllPaintingVarsUnset(String reason, { bool debugDisableShadowsOv
+/// mock client that hasn't been affected by other tests.
+///
+/// This value is ignored in non-debug builds.
+@visibleForTesting
```
why is this one `@visibleForTesting` but not the others?

Seems like it would be reasonable for this one to be usable from non-test debug code too?

tvolkert replied:
I wondered about that - I couldn't think of a concrete use case during debugging.  But happy to remove the annotation - will do.

Hixie commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -68,14 +68,63 @@ class ImageInfo {
+///  * [Image.imageLoadingBuilder], which allows callers to visually represent
+///    the loading progress of an image at the widget layer.
+class ImageChunkEvent extends Diagnosticable {
+  /// Creates a new chunk event.
```
can the arguments be null?

Hixie commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -68,14 +68,63 @@ class ImageInfo {
+  /// This value is not necessarily equal to the expected _size_ of the image
+  /// in bytes, as the bytes required to load the image may be compressed.
+  ///
+  /// This value will be -1 if the number is not known in advance. When this is
```
why -1 rather than null?

tvolkert replied:
'cause it's a passthrough from the `response.contentLength` property, which uses -1.  I could certainly see the argument for making this be null - wdyt?

Hixie replied:
ah if it's just passing through an existing API, sounds good to me.

Hixie replied:
if you haven't already, please reference the api you're forwarding in the docs

Hixie commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -68,14 +68,63 @@ class ImageInfo {
+  ///
+  /// When used with [Image.network], this value comes from the `Content-Length`
+  /// HTTP response header, and as such is only as trustworthy as the server
+  /// that reported it. Callers are correspondingly advised to be defensive
```
Content-Length is always accurate, if present. It has to be, there's no other signal of the end of the resource.

Hixie commented on packages/flutter/lib/src/painting/image_stream.dart:
```diff
@@ -155,11 +213,15 @@ class ImageStream extends Diagnosticable {
   /// was added, even if it was added with a different `onError` than the
   /// intended paired `addListener` call.
-  void addListener(ImageListener listener, { ImageErrorListener onError }) {
+  void addListener(
```
this api is already broken and this makes it more broken. i don't think we should go down this road.

See the docs for removeListener, for example. We'd have to update those if we went this path, and I don't think I could do that without shame.

Hixie commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -705,13 +809,16 @@ class _ImageState extends State<Image> {
       filterQuality: widget.filterQuality,
     );
+    if (widget.imageLoadingBuilder != null && _chunkEvent != null)
+      result = widget.imageLoadingBuilder(context, _chunkEvent, _imageInfo == null ? null : result);
```
we shouldn't create the RawImage for the cases where we don't use it

Hixie commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -660,17 +755,26 @@ class _ImageState extends State<Image> {
-      _imageStream.addListener(_handleImageChanged);
+      _imageStream.addListener(
+        _handleImageChanged,
+        chunkListener: widget.imageLoadingBuilder != null ? _handleImageChunkLoaded : null,
```
as far as i can tell you don't handle widget.imageLoadingBuilder changing over time

Hixie commented on packages/flutter/lib/src/widgets/image.dart:
```diff
@@ -645,10 +732,18 @@ class _ImageState extends State<Image> {
+  void _handleImageChunkLoaded(ImageChunkEvent event) {
+    assert(widget.imageLoadingBuilder != null);
+    setState(() {
+      _chunkEvent = event;
```
this is REALLY expensive. It forces a rebuild for every frame even if the builder doesn't actually return anything that is affected by this information.

tvolkert replied:
But why provide a builder at all if you're not affected by this information?  This information is the essence of why you provided a builder.  And yes, it'll mean that you rebuild every time bytes come across the wire while the image is loading, which is likely every frame.

Hixie replied:
you may wish to show an indeterminate progress indicator while downloading. Or the word "LOADING". Or alternative text in an inline widget.

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -510,25 +517,49 @@ class NetworkImage extends ImageProvider<NetworkImage> {
+    HttpClient client = _sharedHttpClient;
+    assert(() {
+      if (debugNetworkImageUseFreshHttpClient)
+        client = HttpClient();
```
you don't set autoUncompress to false on this one

Hixie commented on packages/flutter/lib/src/painting/image_provider.dart:
```diff
@@ -510,25 +517,49 @@ class NetworkImage extends ImageProvider<NetworkImage> {
-    if (response.statusCode != HttpStatus.ok)
-      throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
+  /// Do not access this field directly; use [_httpClient] instead.
+  static final HttpClient _sharedHttpClient = HttpClient()..autoUncompress = false;
```
you turn off autoUncompress here, but you only support gzip on the other side. HTTP servers can use a variety of compression algorithms other than gzip.

Hixie replied:
(You should make sure to set the `TE` header to whatever it is you support. Also, consider `Accept-Encoding` and `Content-Encoding`, which are different than `TE` and `Transfer-Encoding`.)

tvolkert replied:
HttpClient documents this property as only respecting gzip -- https://api.dartlang.org/stable/2.3.0/dart-io/HttpClient/autoUncompress.html

Hixie commented:
Let's see if we can make FadeInImage a client of this.

tvolkert commented:
I split the `consolidateHttpClientResponseBytes()` changes off into a separate PR - https://github.com/flutter/flutter/pull/32853 since they're isolated.

tvolkert commented:
This is being broken off into multiple sub-PRs.  Here are the ones that have been sent for review thus far:

* https://github.com/flutter/flutter/pull/32853
* https://github.com/flutter/flutter/pull/32857
* https://github.com/flutter/flutter/pull/32936

goderbauer commented:
Should this PR be closed since it split off into multiple new PRs?

tvolkert commented:
Yep, done.
